rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,856 @@
+
+We'll also need to migrate all child modules to a separate monorepo, which we'll call `terraform_modules`. This will allow us to isolate the CI configuration, git tags, and commit history for modules. It will also help reinforce the separation between child modules and root modules, since it will no longer be possible to update both a child module and the infrastructure that uses it in the same pull request.
+
+For details on the alternative module publishing options that were explored, see the [corresponding subsection under rationale and alternatives][alternatives-module-versioning].
+
+### Migration Process
+
+The process to migrate our modules to a versioned and published workflow will be as follows:
+
+1. Create a separate `terraform_modules` repo for private modules
+1. Set up a versioning workflow in GitHub Actions that creates a semver tag when a module is updated
+   - Tags must be named in a manner that [Scalr's module registry supports](https://docs.scalr.io/docs/private-module-registry#sub-modules-mono-repo)
+   - The exact naming convention is to be determined, but will likely be something like `<module-name>/<version>`
+   - Ideally the tagging workflow will be automated and derive the module name and appropriate version string automatically (i.e. from semantic commits or PR tags)
+1. Migrate each infrastructure component and application module to the new repo and rename it to match the required `terraform-aws-<module_name>` convention
+1. Add each module to the private module registry
+1. Update any references in root modules to refer to the module registry, requiring version constraints in production, but not in dev/staging
+
+### Public Vs. Private Registries
+
+There are two options for publishing modules: Using [Terraform's public registry](https://registry.terraform.io/), or using a private registry such as the one [offered by Scalr][scalr-private-registry].
+
+Publishing a module to Terraform's public registry comes with its own [strict set of requirements][terraform-module-publishing], not least of which being that it must be stored in a discrete public repo. Thankfully, private registries is less restrictive on these points, so publishing to a private registry is the preferred approach unless there's a reason to make a module public.
+
+If we do choose to publish a module in Terraform's public registry, the overall process will be similar, except that we'll need to make a separate repo for the module and make other changes in accordance with Terraform's requirements.
+
+### Application Modules In Application Repos?
+
+For application modules, the question has been raised as to whether it might make sense to move the module configuration into the application repo so it lives alongside the codebase. The stated advantages to this approach are as follows:
+
+- All(?) of the application's configuration would live within the same repo
```

Comment:
I don't think it's desirable to tightly couple infrastructure changes to application changes in that way. We should discuss this more as a group outside of this RFC, since it seems to indicate a difference in philosophy around how application infrastructure is managed.

Reply:
I mostly agree, but the way we work now with application configuration specified in Terraform, the coupling is sort of inevitable.

Reply:
I created a task for the tech leads to discuss alternative application configuration approaches.